### ISSUE TITLE
<podcast:location> Web scale IDs

### ISSUE THREAD

**Issue by jamescridland** [Podcastindex-org/podcast-namespace#5]
I like the use of a country code, but would propose to firm that up to a standard.

Some podcasts are "from" a big city; others from a suburb in that city; others from an actual place. A lat/lon doesn't tell us anything about the place being specified, nor its bounding area.

## My suggestion

> **<podcast:location osm_id=[place ID]>**[CountryCode|Locality]**</podcast:location>** - The ISO 3166-1 alpha-2 country code; a pipe as separator; then a humanly-readable place name as preferred by the publisher. The (mandatory) parameter osm_id is the OpenStreetMap ID for that place, using [OpenStreetMap's API](https://nominatim.openstreetmap.org/search.php?q=Columbus%2C%20OH&format=jsonv2).

The use of OpenStreetMap's ID here - OSM is a well-recognised, open and free database - would allow you to programmatically get lat/lons if required; show place names in other languages, grab a rough [bounding box](https://wiki.openstreetmap.org/wiki/Bounding_Box) to aid in searches, and understand the place type. (Use the [osm_id and not the place_id](https://gis.stackexchange.com/questions/133101/difference-between-osm-id-and-place-id)).

[Here is a programmatic call for osm_id 182706](https://nominatim.openstreetmap.org/lookup?osm_ids=R182706&format=json), which includes

```json
[{"osm_id":182706,
"boundingbox":["39.8086936","40.157272","-83.2101797","-82.7713119"],
"lat":"39.9622601","lon":"-83.0007065",
"display_name":"Columbus, Franklin County, Ohio, United States of America",
"class":"boundary","type":"administrative","importance":0.6294399106820546,
"address":{"city":"Columbus","county":"Franklin County",
   "state":"Ohio","country":"United States of America","country_code":"us"}}]
```

## Valid examples

`<podcast:location osm_id="182706">[US|Columbus, Ohio]</podcast:location>`
`<podcast:location osm_id="182706">[US|Columbus, OH]</podcast:location>`
(These are both the same place)

`<podcast:location osm_id="1850813">[US|Columbus]</podcast:location>`
(This is actually Columbus in Georgia)

`<podcast:location osm_id="65606">[GB|London]</podcast:location>`
`<podcast:location osm_id="7646215719">[GB|Westminster, London]</podcast:location>`
`<podcast:location osm_id="1567699">[GB|Houses of Parliament, London]</podcast:location>`
The above might all describe the same podcast, but there's significant benefit to use the last one. You might know it as the Houses of Parliament, but the correct name, as OSM will tell you, is the "Palace of Westminster". The "bounding box" all of these gets progressively smaller, so a podcast for "London" is very different for one just for "Westminster".

`<podcast:location osm_id="175905">[US|New York]</podcast:location>`
`<podcast:location osm_id="8398124">[US|Manhattan]</podcast:location>`
The first one is actually New York City, rather than New York State, which can often be confused.
But what if the podcast is actually about Manhattan, and not Brooklyn?
For the NYC listing, it has `"boundingbox":["40.477399","40.9161785","-74.25909","-73.7001809"]` which covers all of New York City. But for Manhattan, you can get `"boundingbox":["40.6839411","40.8804489","-74.0472219","-73.9061585"]` which is just the Manhattan area.

`<podcast:location osm_id="65606">[GB|London]</podcast:location>`
An app configured in French can programmatically [grab the right name in that language](https://nominatim.openstreetmap.org/lookup?osm_ids=R65606&format=json&accept-language=fr) and return "Londres, Angleterre, Royaume-Uni"

## Data requirements

OSM is open. You can (and should) maintain your own database, but there are plenty of mirrors around the world. It does have a licence which requires credit. It would be helpful to understand what the deal is here with RSS feeds. I'm a member of the OSM Foundation, so I could go and peek a little more. Alternatives are Google Maps (proprietary), or [GeoNames](https://www.geonames.org/) which still has an attribution requirement, and looks to be relatively un-maintained judging by the traffic in its forum.

Hope this shows the benefits of not just using raw text. Let's do this properly! ;)

**Comment by daveajones** [comment]
Yes.  This is a biggie.  Let me write this in as best I can to represent what you have here and you can look it over and give me your feedback.

**Comment by jamescridland** [comment]
Thanks. I'm being bullish and asking for the `osm_id` to be mandatory. You might not agree with that.

(I guess I should make it clear that the additional data you can get from OSM probably shouldn't be added into the RSS. First, it might change - the Houses of Parliament in Australia have moved at least once, as one example, and the 'bounding_box' almost certainly will change in time as places expand or contract. And second, my intention is for podcast indexes and search engines (like podcastindex.org) to be _able to_ incorporate that into their search engines or services. The end-client, like a podcast app, is unlikely to benefit from making any calls to OSM itself, but could certainly parse and display the "NL|Amsterdam" value.)

As one example of how this might work for search:

"Podcasts from near me" - "AU|Brisbane" is of no particular use to a search engine algorithm, other than simply doing a text search. With a lat/lon you're able to suggest podcasts that come from places that are 10km or less away from me, so you'd also catch "AU|Milton" and "AU|Bardon". And with a 'bounding box', you're able to realise that while I (I'm the "O" in this) am nice and close to both Milton and Brisbane, the correct order would/could be "Bardon", then "Brisbane" (because I'm in it), then "Milton" (because I'm not).

![Screen Shot 2020-10-10 at 7 08 44 pm](https://user-images.githubusercontent.com/231941/95651208-0711b680-0b2c-11eb-81f1-6947c01197ea.png)
"O" = me (kind of)
"X" = lat/lon point for each

**Comment by jamescridland** [comment]
Sorry. One additional thought, as I write to OSM...

This can live in the CHANNEL but also in the ITEM, right?

A travel podcast might have one show in Washington DC, and then the next show focusing on Boston MA.

**Comment by daveajones** [comment]
Yes absolutely this is channel level and item level.

If you think osm_id is mandatory, then the next question would be is the location tag itself mandatory if someone implements the namespace. I think it should be.

Podcast hosting services should allow their subscribers to set their location by searching their locality by name and then showing them the OSM map with the bounding box. That is then

**Comment by saerdnaer** [comment]
@jamescridland Please do not assume that an OSM node_id is permanent – you have to at least include the version number or better even the coordinates, e.g. https://wiki.openstreetmap.org/wiki/Permanent_ID#OSM_Permanent_ID_with_geo_URI

**Comment by daveajones** [comment]
I've incorporated your changes @jamescridland .  I'll wait on any followup comments you have to @saerdnaer before moving forward with any more changes to this element.

**Comment by jamescridland** [comment]
@daveajones - I'm always hesitant about making a tag mandatory. There may be good reason not to set a podcast:location at all. I'd appreciate other views here. I do think it would be helpful that *if* the tag is used, *then* the osm permalink is mandatory - otherwise this will be relatively useless for actual use - a structured data field filled with just raw text is a bad thing.

> For existing customers the initial values upon implementation could just come from their existing customer location data and generalized into a very broad bounding box for privacy.

Not a bad idea. This might simply be the OSM id for "USA", and the entire bounding box of the country: but at least that would be a start, and more that we have now. (Which does mean that we should allow the "locality" to be optional - so **<podcast:location osm_id=[place ID]>US</podcast:location>** is acceptable.

> Are you contacting the OSM to see if referencing their ids in the spec is kosher with their license?

I have done. I'll wait to see when they come back.

@saerdnaer Thank you for catching that. Goodness, that's frustrating - but at least we found it now rather than later!

The suggested permanent ID of "geo:47.45835,8.74375?q=relation/1169711#6" looks doable, I suppose - at least that has the possibility that even if an ID disappears (unlikely for a city, more likely for "Houses of Parliament") then we at least have a lat/lon. That is, though, just a proposal.

We *could* have this construct:

> <podcast:location latlon:"47.45835,8.74375" osmid="R1169711#6">US|Springfield, OH</podcast:location>

... which has the same data in it, but allows for easier parsing?

**Comment by jamescridland** [comment]
So - todo here is:

1. Feedback from OSM. If their licence doesn't allow this, then it kills it.

2. Are both `latlon` and `osmid` mandatory if the podcast:location tag is used? I would like to propose that they are.

3. Is podcast:location a mandatory tag? I could be convinced either way.

(To be honest - having lat/lon in the data does allow apps to do some nice UX without a call to OSM, which just having a raw ID precludes - so this seems a better solution all round)

**Comment by daveajones** [comment]
1. 10-4. Will wait on you.

2. I’m fine with this.

3. I think it can be. As a new spec, this will just be a part of the code we all have to write in the next month or so. All the industry participants that are willing to code for this namespace will also see this tag’s value from day one. The ones who aren’t willing to code for this tag probably don’t care about playing ball with the needs of the industry anyway.

I consider this tag, along with “locked”, the “email” verification tag and the “alternateEnclosure” tag to be critical functionality of this namespace.

In that regard, maybe only “latlong” is required?  It give us what we need and makes implementation on the hosting company side easier if they already support google maps widgets or some other library in their GUI.

**Comment by jamescridland** [comment]
I could be persuaded that we go with a version where latlon and country code is mandatory, and the osmid (and locality) is recommended?

I'd really like to see the OSMID in there for the richness of data it gives, but I can also see that implementation would be significantly easier without it. At the very least, this leaves us with a) a really simple standardised country code; b) lat/lon for a simple map; c) an optional podcaster-led description in plain text of the location for display.

@saerdnaer - The lack of full permanence for an OSMID isn't, I don't feel, an impediment. Cities and major placenames are unlikely to disappear; and no geographic naming system will ever give permanent IDs. The OSM version (the portion after the hash) is not programmatically available in the OSM APIs, and nor can it be specified for API calls anyway. I'm happy to keep digging here though, and am currently trying to find data about the permanence of osmid for placenames.

How does this look for a rather wordy definition?

----
* **<podcast:location latlon:"[latitude,longitude]" (osmid="[OSM type][OSM id]")>**[CountryCode(|Locality)]**</podcast:location>**` - this is a mandatory channel-level element (and an optional item-level elelement). Podcasts MUST contain a latlon point and a country code.
  * latlon: **mandatory:** a latitude/longitude point reflecting the location associated with this show or episode. This could be where it is made, or alternatively a location which features in the podcast.
  * osmid: **recommended** - from [the OpenStreetMap API](https://nominatim.openstreetmap.org/search.php?q=Columbus%2C%20OH&format=jsonv2). If a value is given for osmid it must contain both 'type' and 'id'. 
    * osm type: a one-character description of the type of OSM point. Valid is "N" (node); "W" (way); "R" (relation).
    * osm id: the ID of the OpenStreetMap feature that is described. This may be a city or a building. While OSM IDs are not considered permanent, cities rarely disappear.
  * CountryCode: **mandatory** the ISO 3166-1 alpha-2 country code, eg 'US'. (Note that the United Kingdom is 'GB', not 'UK'.)
  * Locality: **recommended** - with a pipe separator from the countrycode, this is a humanly-readable place name as preferred by the podcast publisher.

Mandatory level of detail:
> <podcast:location latlon:"39.96,83.00">US</podcast:location>

Recommended level of detail:
> <podcast:location latlon:"39.9622,83.0007" osm_id="R182706">US|Columbus, OH</podcast:location>
> <podcast:location latlon:"51.49942,-0.12444" osm_id="R1567699">GB|Houses of Parliament, London</podcast:location>

----
Questions:

a) if the user simply just wants to say "My podcast is from Iceland", what should the lat/lon point to? The capital of the country? The middle of the landmass? Or do we not care?

b) do we have a requirement for the accuracy of the lat/lon? [This document suggests](https://gisjames.wordpress.com/2016/04/27/deciding-how-many-decimal-places-to-include-when-reporting-latitude-and-longitude/) that a latlon like "39.96,83.00" is a kilometer wide at the equator; "39.962,83.000" is 111m wide (ten times more accurate). Perhaps we should be suggesting "for most places, three decimal places in the lat/lon should be sufficient"?

**Comment by tomrossi7** [comment]
I like the idea of making it optional, but I'm hesitant to make location data required. From the podcaster's perspective, that could be invasive if they don't necessarily think its relevant to their podcast.

**Comment by daveajones** [comment]
> 
> 
> I like the idea of making it optional, but I'm hesitant to make location data required. From the podcaster's perspective, that could be invasive if they don't necessarily think its relevant to their podcast.

When you say: "hesitant to make location data required" are you talking about the whole tag, or just the latlon part?

The latlon point and country code can be made very non-specific.  For example, the country code could be "US" and the latlon just a point at the center of the US.  Like this:  https://www.openstreetmap.org/relation/148838#map=4/40.55/-50.19

But, I see your point.  Some people could be very privacy conscious in that regard.

The point of this tag is an opportunity to give apps very rich data about local relevance.  Being able to recommend other local podcasts without resorting to big data would be joyous.

What about keeping the tag required and just allowing for this:

`<podcast:location private>`

... which could map directly to a "Keep my location private" check-box in the podcaster's dashboard?

@jamescridland Thoughts?

**Comment by jamescridland** [comment]
@daveajones I think this is a good pragmatic compromise.

Making **<podcast:location>** mandatory, but allowing people to just mark it as "private", I think neatly both ensures implementation by publishers, and an opt-out for those that might just fill it with unhelpful junk.

I haven't heard from OSM yet regarding their licence requirement here. I'll give them a prod to a wider email list.

**Comment by jamescridland** [comment]
Having failed to get a response from the OSM legal questions team, I'm now asking [in the OSM help forum](https://help.openstreetmap.org/questions/77084/license-requirement-for-use-in-a-new-podcast-tag) about attribution.

**Comment by jamescridland** [comment]
OK: while the OSM help forum has been useless too, [this page](https://wiki.osmfoundation.org/wiki/Licence/Community_Guidelines/Geocoding_-_Guideline) has the following example, based on a database of store locations:

> A retail store operator uses the Geocoder to populate lat/long locations for certain locations in their database of store locations (Forward Geocoding). The geo-services vendor is required to credit OpenStreetMap as described in Section 4.3 of the ODbL because it is Publicly Using the OSM database. The Geocoded Results added to the store location database are not subject to the share-alike requirements, and the store location database records themselves need not include attribution to OpenStreetMap because the Geocoding Results used are an insubstantial extract or contain no OSM data.

I would think this is the equivalent of:

> A podcast host uses the Geocoder to populate lat/long locations for podcasts in their database of hosted shows (Forward Geocoding). The podcast host is required to credit OpenStreetMap as described in Section 4.3 of the ODbL because it is Publicly Using the OSM database.

...i.e if you're a podcast host using OSM, you need to probably say so somewhere. This is outside our scope. And they can use anything else they want too (Wikipedia, Google Maps, etc).

> The Geocoded Results added to the RSS feeds are not subject to the share-alike requirements, and the podcast location database records themselves need not include attribution to OpenStreetMap because the Geocoding Results used are an insubstantial extract or contain no OSM data.

I'm therefore comfortable that OSM ids do not require attribution, though programmatic use of them would.

So that's nice.

[This page](https://wiki.openstreetmap.org/wiki/Wikidata#OpenStreetMap-related_properties_at_Wikidata) talks about Wikidata linking, and points out that "Ways and nodes are not considered stable IDs by Wikidata" - and that Wikidata only accepts Relational IDs (the "R" in the above example). It would seem sensible to mirror this - and that we only link to relational OSM ids.

So, here's where we currently are:

----

* **<podcast:location latlon:"[latitude,longitude]" (osmid="[OSM type][OSM id]")>**[CountryCode(|Locality)]**</podcast:location>** - this is a mandatory channel-level element (and an optional item-level elelement). Podcasts MUST contain a latlon point and a country code.
  * latlon: **mandatory:** a latitude/longitude point reflecting the location associated with this show or episode. This could be where it is made, or alternatively a location which features in the podcast.
  * osmid: **recommended** - from [the OpenStreetMap API](https://nominatim.openstreetmap.org/search.php?q=Columbus%2C%20OH&format=jsonv2). If a value is given for osmid it must contain both 'type' and 'id'. 
    * osm type: a one-character description of the type of OSM point. The only valid type currently is "R" (relation); nodes and ways are [not considered permanent enough](https://wiki.openstreetmap.org/wiki/Wikidata#OpenStreetMap-related_properties_at_Wikidata) by Wikidata.
    * osm id: the ID of the OpenStreetMap feature that is described. This may be a city or a building.
  * CountryCode: **mandatory** the ISO 3166-1 alpha-2 country code, eg 'US'. (Note that the United Kingdom is 'GB', not 'UK'.)
  * Locality: **recommended** - with a pipe separator after the countrycode, this is a humanly-readable place name as preferred by the podcast publisher.

Mandatory level of detail:
> <podcast:location latlon:"39.96,83.00">US</podcast:location>

Recommended level of detail:
> <podcast:location latlon:"39.9622,83.0007" osmid="R182706">US|Columbus, OH</podcast:location>
> <podcast:location latlon:"51.49942,-0.12444" osmid="R1567699">GB|Houses of Parliament, London</podcast:location>

----

After all this, I do slightly worry that nobody will bother with the OSM bit. But I still think it's worth battling on for!

**Comment by daveajones** [comment]
I agree James. It’s worth it to have it available for use even if underutilized.  What you’re showing seems like a great place to land on this. 

I’ll make these changes today.  And, we can consider this one fully baked for now.

**Comment by jamescridland** [comment]
(Sorry - just re-opening this to apologise for a typo in the spec, which I'm sure you'd pick up on anyway:

latlon:"[latitude,longitude]"

should read

latlon="[latitude,longitude]"

**Comment by tomrossi7** [comment]
Just my 2 cents, but having it required and marked `private` adds unnecessary tags to the RSS. I would prefer to keep it optional and let the platform and podcaster decide whether to provide it. Even if Buzzsprout wanted to support the tag, we wouldn't have the underlying data and would need to mark hundreds of thousands of RSS feeds `private` .

**Comment by jamescridland** [comment]
I think @tomrossi7 makes a good point. "private" should only be there if the podcaster wants to make it private, really - and the point about podcast hosts not having this data is fair enough. Dropping it down to optional is, I guess, OK. So, perhaps we don't need the `private` tag at all, to keep things simpler.

With all these tags, I think it important to highlight the good that will come out of them.

"This is an optional channel or item element. Its purpose is to help discovery of podcasts containing local content."

**Comment by daveajones** [comment]
Ok I’ll make this change and we’ll let it simmer for a bit. Yes, I think the adoption can be driven by marketing. Podcastindex can put up a cool “find your local podcasts” map where only the ones defining the tag show up. Something like that to show the need.

**Comment by tomrossi7** [comment]
`"This is an optional channel or item element. Its purpose is to help discovery of podcasts containing local content."`

This is exciting! The reason we are undertaking this project is to extend whats possible with RSS as it exists today. This is a great example of how we can share information between hosts and aggregators to accomplish *new* functionality - a platform for innovation rather than just an updated specification.

**Comment by daveajones** [comment]
100%

**Comment by douglaskastle** [comment]
I don't understand why the country code is mandatory.  To be of any use the podcatcher or the website has to be quiet smart to deliver suggestions i.e. be aware of the region is influence that a country code occupies, (however i know that some tools like django/GeoDjango provide this fairly easily). 

Surely a latlon and a radius of validity would be more useful. Search functions become a lot more simpler? It becomes a function of basic geometry.  It would still be up to the podcaster or hosting service to determine the correct latlon and radius.

**Comment by jamescridland** [comment]
As a programmer - it is much, much easier to match on "FR" than to look up every single lat/long to work out whether it falls in France, Belgium, Germany, Luxembourg, Spain, Monaco or even the UK.

Most IP geo-databases resolve down to a country code, so it's also simple to match to show a user other podcasts from their country. Some may also use device language, which isn't perfect but would also allow a match with a country.

It is also trivial to add as a podcast host.

**Comment by douglaskastle** [comment]
It is the mandatory I have an issue with.  There is no way to guarantee that the country code matches the latlon. (though there is no way to guarantee that the latlon is even right, but lets just assume from here on that it is)

Would it not make sense for what ever search tool you use, it is the thing that as a first step resolves the latlon down to something that can be used. So either a geometric search by radius or country code, which is just regex and can be a bit blunt. Getting the user to add the latlon AND the country code gets a bit tedious and ultimately redundant.

There are spatial databases that are designed to do latlon searches, and they can be country/border aware.  

What is the concern with the FR example, versus all the other countries that border it. Brussels is closer to Paris than Perpignan is, what is more of interest when you are search by location.  Again this I think comes back to a radius of influence more so.  Content for a tourist spot might be of more localized interest than a news report about a country (for example).

Again, as an available field it may be of use for some lazy, underpowered search tools, I am here thinking about a mobile app podcatcher search feature, but making it mandatory seems wrong to me. The real power of this tag is being able to deploy a full featured (and populated) spatial databases.

**Comment by jamescridland** [comment]
> Getting the user to add the latlon AND the country code

I don't think anyone is requesting that this is done manually. When a user clicks on a map or enters an address, a latlon *and* a country code will be produced by the geotagging application used in any case. There's no additional work on the user's behalf, nor on the podcast host's behalf to populate both these fields.

The country code is a function of the lat/lon. It allows a simple method of calculating "Which country is this podcast from/about".

> The real power of this tag is being able to deploy a full featured (and populated) spatial databases.

Agree; but if we are to ONLY have lat/lon, then it makes it significantly harder to do simpler searches and filters.

In the below - which are in France?

48.8566, 2.3522
50.8503, 4.3517
50.6292, 3.0573

Without looking up each and every latlon in a database, you cannot discover this information. This might be trivial for this example; but when there are 1.3m podcasts out there, it's not simple to produce something which makes 1.3m geo-database lookups just to discover a country code.

To use an ISO standard country code as a mandatory field:
* allows simple client-side filter/search for podcasts from a specific country
* is no additional work for anyone, since this data will be available from any geocoder
* makes life simpler and easier for developers

> Brussels is closer to Paris than Perpignan is, what is more of interest when you are search by location.

a - "Podcasts from my country" might be desired.
b - Since this is not a binary choice, the latlon still allows you to calculate a "radius of influence".

There are significant benefits of having the country code in this tag (and, even more benefit to have an OSM ID, since that includes bounding boxes and other things that assist a more advanced search). But I'd not be too keen to remove the country code, given it makes things easy for "lazy, underpowered search tools", while not adding any additional hard work for producers.

**Comment by douglaskastle** [comment]
I respect the analysis and while I certainly can't tell you which latlon's are in France, this is exactly a case where one would use a computer to resolve that.  I have to ask, are you familiar with how a spatial database works, it is quiet  literally designed to do exactly that type of search, by design, maybe you just feel computationally it is expensive?

[Spatial Database](https://en.wikipedia.org/wiki/Spatial_database)

But what you haven't answered, I feel, is why does the podcast tag have to carry the country code as a mandatory requirement.  Why can't the search tool just resolve the latlon to a country code as it first step.  Having it in the search tool feels like the correct place for this to happen.  It should carry the locationally aware smarts. If the search tool uses country codes to do search well cool.  If it doesn't and uses a full spatial database search it is redundant.  But it doesn't matter, the search tool manages what it wants from the latlon. 

Also country codes may change down the road, 5-10 years, South Sudan and North Macedonia didn't exist 10 years ago (for example, though I don't know if country codes were changed as a result). The podcasts will not be updated in the same way as the search tools are.

So I am not saying don't use country codes, i just don't feel making them mandatory in the podcast tag makes long term sense, but by all means keep it as an optional value.

**Comment by jamescridland** [comment]
>while I certainly can't tell you which latlon's are in France, this is exactly a case where one would use a computer to resolve that

Yeah, and I think by asking for the country to be mandatory data, it allows this to be done very easily without inspecting every single lat/lon data point - i.e. without the need for a spatial database. As someone who maps for OpenStreetMaps every day, I know how difficult some of this data can be!

> Why can't the search tool just resolve the latlon to a country code as it first step?

It can. It can absolutely do that. Nothing in asking for the country code to be mandatory stops this from happening. I would absolutely agree that latlons are mandatory. They have to be. In any case, this is a *much* better plan than having an open text field saying "London" in it.

The country codes are there to make it easy to produce one commonly-requested filter - "podcasts from x country" - **without** requiring the podcast app to have access to a full database of which countries lie in which latlons. It's making life easier for podcast developers with no negative impact.

You're absolutely right that `SS` and `MK` were not ISO-3166 country codes ten years ago. However, Barack Obama is no longer the US President, but there are plenty of podcasts out there recorded when he was. Podcast data is historical by its very nature.

So - the point of putting the country-code as a mandatory field is:
* it's really simple to do so, and doesn't add any extra work for anyone
* it allows simple filtering of podcasts as a result with no latlon lookups
* it doesn't get in the way of anything else and makes developers' lives easier to achieve an obvious filter
* including it as mandatory will not harm the takeup of this tag
* it's really hard to stuff up a country code

Making them optional means that the simple option of the filter won't work since the data won't be as complete. It doesn't remove any work for the podcast host (which still needs to calculate the latlons), and makes life appreciably harder for a developer. I want to make this as rich data as possible, hence the OSM links which give much more rich information than a point on the planet; and as simple for a developer to start using as possible - hence the simple country code.

I don't quite understand the fierce pushback, since I don't quite understand what issues it causes by being mandatory; but I'm grateful for the opportunity to clarify.

**Comment by douglaskastle** [comment]
> I don't quite understand the fierce pushback, since I don't quite understand what issues it causes by being mandatory; but I'm grateful for the opportunity to clarify.

So for context my day job is a microchip designer.  Our work is finished in baked sand.  Virtually impossible to fix later.  So it is important to have all the discussions at the start in detail, because you can't have them later.

Software is more forgiving, however if there was ever an area that is close to the problem of chip design, it is standards.  Getting them set up is one thing, changing them later can be next to impossible.  Once out in the field you get a lot of push back in cases where it has been implemented already.  So again, here and now is where to have these conversations up front and robust!

I hear all you arguments for search.  But it is a backwards looking argument. You are trying to figure out how to bolt on the new functionality with existing infrastructure.  The **_correct_** place for this is on the tool doing the search, not the tool doing the recording of the tag.  The search tool can be changed and updated later, the podcast will never change, so some (but not all) of that information is going stale, and in that case it would be ignored, so you are back to the same situation.

If a developer can't figure out how to resolve a country code from a latlon, then maybe they shouldn't be trying to provide a geo aware search algorithm.  If they can do that, they can do everything you suggested, why is this viewed as such heavy lifting for the developer of the search and not heavy lifting for the tool that helps record it in the podcast geotag.  Maybe there needs to be an interim service to help them return a country code from a latlon?

Look, I am not looking for a fight, I just don't think it should be made mandatory as part of the spec, I am not saying don't leave the field there to be used.  I am prototyping a geo aware podcast search, I won't be using the country codes, or if I ever do I would just derive them straight form the latlon. Also I suspect this is something that will actually just be ignored too if it ever does get implemented, standards there too are just standards and are regularly violated.

**Comment by jamescridland** [comment]
Douglas - thanks. I still don't understand the concern.

> If a developer can't figure out how to resolve a country code from a latlon, then maybe they shouldn't be trying to provide a geo aware search algorithm.

This is a bit of a condescending attitude, Douglas. It isn't for lack of knowledge - it's to make things easier for one simple filter that is often requested. It isn't just for search, either - another UX element might be a simple country flag for the show; or to simply filter a list of shows returned.

The fact is that it is non-trivial to do a full lookup for 1.3m different items in a thin podcast client. However, it *is* trivial to add this information into the RSS feed in the first place.

> why is this viewed as such heavy lifting for the developer of the search and not heavy lifting for the tool that helps record it in the podcast geotag

Because in order to discover a latlon, you are - in most cases - automatically presented with the country code by the geocoder you use. [Like this one](https://nominatim.openstreetmap.org/lookup?osm_ids=R182706&format=json). For the podcast publisher, there is literally nothing extra to do to populate this field. This is easy, simple, and adds no extra work for anyone. The same simply cannot be said for looking up 1.3m latlons to discover countries.

> I am prototyping a geo aware podcast search, I won't be using the country codes, or if I ever do I would just derive them straight form the latlon

Awesome. Delighted that is good for you. Choice is a good thing. Let's retain that choice.

I'd have preferred to have only used OSM IDs, and make those mandatory (as you'll spot from the initial proposal). However, I recognise that, pragmatically, it isn't brilliant to do 1.3m OSM lookups either. At least the mandatory country code may spur some on to completing the OSM data as well - which would be MUCH more useful.

You're welcome not to use the country code, and in many cases, using latlons (or, far better, OSM IDs) will give a better outcome. But for the simplest of all cases, there is no need to give a developer more work to do to just display a country flag, or a simple filter on "show me podcasts from Germany only".

> Also I suspect this is something that will actually just be ignored too, if it ever does get implemented, standards there too are just standards and are regularly violated.

Indeed. But this proposal adds literally no additional complication for a publisher, while actively making life easier for some use-cases for a client. If the only objection from you is "But I'm not going to use it", then that's cool.

We do appear to be going round in circles, and this proposal is closed anyway. I haven't yet been convinced of any reason to drop the mandatory-ness of the country code. I'm very happy to be convinced, but a better reason than "other developers are stupid" or "I won't use it" would be really required here.

**Comment by douglaskastle** [comment]
> This is a bit of a condescending attitude, Douglas.

I am sorry it is not.  In business it is routinely required for people to define and assess minimum level of competency to implement features. Geo-search in podcasting is relatively unheard of.  I would have no problem placing that requirement on my team or have it expected of me, if this feature was to be implemented into an app.  You may want to use it yourself if you are hiring talent for [maps.fm](https://maps.fm/)

> this proposal is closed anyway

This is condescending. It is the equivalent of I have taken my ball home and you can't play.  Issues can be reopened. It is unfortunate that I missed the start of the conversation, I was only made aware of it after it was mentioned on a recent podcasting 2.0 podcast.  There, they said, come here to have the conversation.

Let me finish with this.  You have made no case for why it should be made mandatory in a standard.  You have only made a case for why you want it mandatory based on some use case you have in mind or are already implementing.  This is called feature creep, having the tail wag the dog. This happens in products all the time, this should not happen in standards (if possible). Note I have also not said get rid of it, leave it as a recommended.

- There is no extra information in a country code that cannot not derived from the latlon
- The country code can go stale, so it can contain less information than the latlon
- the country code can be wrong and not actually match the latlon.

What you are missing is that there needs to be a database, somewhere, of searchable meta data, no thin client can be expected to do the heavy lifting of search.  Here is where you would store all your latlons/country codes (that should happen if it is mandatory or not).  This will be true for all the other namespace additions, `person `is a big one I think that will benefit from this.

**Comment by jamescridland** [comment]
The proposal was closed by Dave Jones.

I'll leave it up to him as to whether he wishes to reopen it, but I always prefer constructive discussion.

**Comment by daveajones** [comment]
I have no problem re-opening this. Just want to make sure people don’t have to read 20 pages of material to understand the issue at hand.

I think it’s important to always acknowledge that “required“ is a little bit of a fantasy when it comes to name spaces. Realistically, people are going to do what they want. And, they will also make mistakes.  It’s the old adage of be conservative with what you send, and liberal with what you accept.

@douglaskastle I did my best to read through everything that was said, but for clarity are you saying that, in your opinion, the two-letter country code should not be required but the latitude/longitude should be?

**Comment by douglaskastle** [comment]
@daveajones Yes the latitude/longitude should be mandatory. I am only arguing that the country code, while useful and should be available as a field, I don't think a dependency should be placed on it to be mandatory, from a standards perspective.

**Comment by swschilke** [comment]
Hallo, just wanted to mention that not only location but also language should be available in a tag. Either with this or as a separate tag (I opened an issue #107

**Comment by douglaskastle** [comment]
I don't think language should be captured in the location tag, they are separate things.  I do think it is important, so moving that conversation to #107

**Comment by benjaminbellamy** [comment]
Hi all,

If I may add my 2 cents here:
I also think that country code should not be mandatory (although it will probably be there for 99%).

Because country is not necessarily what matters:
- what is important is not where the podcast was made but what it is talking about
- a podcast can be about many countries
- there are many topics that cross borders
- many OSM object are on multiple countries (eg. Europe: OSMID=25871341, The Alps: OSMID=2698607…)
- the Nominatim API (which I guess everyone will use to fetch OSM ids) does not return a country code

Ben.

**Comment by daveajones** [comment]
I think I ultimately fall to keeping country required, for one simple reason:  complexity.  If you make a standard that's too hard to interop with, devs will ignore it.  Latlon/OSM may not be "hard" to resolve in and of themselves.  But, compared to simply recording a country code in your database at lightning speed during aggregation, they _are_ way harder.  An API lookup will always be more complicated/fragile than simply recording a value.  I think this is what @jamescridland is trying to express with:

```
So - the point of putting the country-code as a mandatory field is:

   - it's really simple to do so, and doesn't add any extra work for anyone
   - it allows simple filtering of podcasts as a result with no latlon lookups
   - it doesn't get in the way of anything else and makes developers' lives easier to achieve an obvious filter
   - including it as mandatory will not harm the takeup of this tag
   - it's really hard to stuff up a country code
```

@douglaskastle and @benjaminbellamy have good points that I think we can solve by simply stating that the country code **is** required to be present in the tag - **but**, if latlon/osm is present, then what they resolve to takes priority.  This allows hosters/generators creating the feeds to always have a solid, simple standard to start with that can be overridden by a more robust solution when/if they choose it.

Basically, it's required, but not always authoritative if superseding data is also present.

Could we move forward with this solution?

**Comment by douglaskastle** [comment]
That's OK with me, thank you for reopening and reconsidering my proposal.

**Comment by jamescridland** [comment]
Works for me too. A sensible solution.

**Comment by benjaminbellamy** [comment]
OK for me.
